InvenTree/plugin/builtin/integration/mixins.py

```diff
@@ -180,6 +181,69 @@ def unregister_tasks(self):
             logger.warning("unregister_tasks failed, database not ready")
 
 
+class EventMixin:
+    """
+    Mixin that provides support for responding to triggered events.
+
+    Implementing classes must provide a list of tuples,
+    which provide pairs of 'event':'function'
+
+    Notes:
+
+    Events are called by name, and based on the django signal nomenclature,
+    e.g. 'part.pre_save'
+
+    Receiving functions must be prototyped to match the 'event' they receive.
+
+    Example:
+
+    EVENTS = [
+        ('part.pre_save', 'myplugin.functions.do_stuff'),
+        ('build.complete', 'myplugin.functions.process_completed_build'),
+    ]
+    """
+
+    # Override this in subclass model
+    EVENTS = []
+
+    class MixinMeta:
+        MIXIN_NAME = 'Events'
+
+    def __init__(self):
+        super().__init__()
+        self.add_mixin('events', 'has_events', __class__)
+        self.events = getattr(self, 'EVENTS', [])
```

Comment:
It could be beneficial to use an (overridable) function to get the EVENTS variable so that a plugin could generate events dynamically or pass function references to the currently initialized instance of the plugin to the events (so that the plugin can have a state-machine / states and respect them when processing events). Totally not a must.

Reply:
check L260 - 264 for an example

Reply:
Good idea, this would allow events to be generated at 'run-time' based on configured settings.

Reply:
Yes, that is kind of the idea. I wrote the plugin mixin code to be basically just a bunch of functions you can override to change everything. This is why the activation sequences are in the registry.py so they can not be mangled with and permissions to execute are respected.